lib/srv/db/sqlserver/kinit/kinit.go

```diff
@@ -0,0 +1,314 @@
+		userName:        user,
+		cacheName:       fmt.Sprintf("%s@%s", user, realm),
+		RealmName:       realm,
+		KDCHostName:     kdcHost,
+		AdminServerName: adminServer,
+		dataDir:         dataDir,
+		certPath:        fmt.Sprintf("%s.pem", user),
+		keyPath:         fmt.Sprintf("%s-key.pem", user),
+		binary:          kinitBinary,
+		command:         command,
+		certGetter:      certGetter,
+		ldapCertificate: ldapCA,
+		log:             logrus.StandardLogger(),
+	}
+}
+
+// commandGenerator is a small interface for wrapping *exec.Cmd
+type commandGenerator interface {
+	// CommandContext is a wrapper for creating a command
+	CommandContext(ctx context.Context, name string, args ...string) *exec.Cmd
+}
+
+// execCmd is a small wrapper around exec.Cmd
+type execCmd struct {
+}
+
+// CommandContext returns exec.CommandContext
+func (*execCmd) CommandContext(ctx context.Context, name string, args ...string) *exec.Cmd {
+	return exec.CommandContext(ctx, name, args...)
+}
+
+// CommandLineInitializer uses a command line `kinit` binary to provide a kerberos CCache
+type CommandLineInitializer struct {
+	auth auth.ClientI
+
+	// RealmName is the kerberos realm name (domain name, like `example.com`
+	RealmName string
+	// KDCHostName is the key distribution center host name (usually AD host, like ad.example.com)
+	KDCHostName string
+	// AdminServerName is the admin server name (usually AD host)
+	AdminServerName string
+
+	dataDir   string
+	userName  string
+	cacheName string
+
+	certPath string
+	keyPath  string
+	binary   string
+
+	command    commandGenerator
+	certGetter CertGetter
+
+	ldapCertificate *x509.Certificate
+	log             logrus.FieldLogger
+}
+
+// CertGetter is an interface for getting a new cert/key pair along with a CA cert
+type CertGetter interface {
+	// GetCertificateBytes returns a new cert/key pair along with a CA for use with x509 Auth
+	GetCertificateBytes(ctx context.Context) (certPEM, keyPEM, caCert []byte, err error)
+}
+
+// DBCertGetter obtains a new cert/key pair along with the Teleport database CA
+type DBCertGetter struct {
+	// Auth is the auth client
+	Auth auth.ClientI
+	// KDCHostName is the name of the key distribution center host
+	KDCHostName string
+	// RealmName is the kerberos realm name (domain name)
+	RealmName string
+	// AdminServerName is the name of the admin server. Usually same as the KDC
+	AdminServerName string
+	// UserName is the database username
+	UserName string
+	// LDAPCA is the windows ldap certificate
+	LDAPCA *x509.Certificate
+}
+
+// GetCertificateBytes returns a new cert/key pem and the DB CA bytes
+func (d *DBCertGetter) GetCertificateBytes(ctx context.Context) (certPEM, keyPEM, caCert []byte, err error) {
+	clusterName, err := d.Auth.GetClusterName()
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+
+	certPEM, keyPEM, err = windows.CertKeyPEM(ctx, d.UserName, d.RealmName, time.Second*60*60, clusterName.GetClusterName(), windows.LDAPConfig{
```

Comment:
Do we really want to generate Windows certs ? inside the `windows.CertKeyPEM` flow the `authClient.GenerateWindowsDesktopCert` auth function is called. Auth service allows to call this endpoint by identities with [types.RoleWindowsDesktop](https://github.com/gravitational/teleport/blob/962e5a25bc56b5aad05528867182a43cd53cb551/lib/auth/auth_with_roles.go#L5046) role where db agent has `RoleDatabase` role.

Reply:
I don't see an easy way around it; we're going to need to generate these certificates with the windows specific extensions or the windows AD/KDC just won't use them to verify identity on their side when we initiate the kerberos connections. I could also develop an entirely separate endpoint that mostly copies what's already there but is only for SQLServer.

In my final version of this I have, which I think makes sense to move here, but we may need to discuss this further with @zmb3 
```
// GenerateWindowsDesktopCert generates a certificate for Windows RDP or SQL Server
// authentication.
func (a *ServerWithRoles) GenerateWindowsDesktopCert(ctx context.Context, req *proto.WindowsDesktopCertRequest) (*proto.WindowsDesktopCertResponse, error) {
	// Only windows_desktop_service or database_service should be requesting Windows certificates.
	if !a.hasBuiltinRole(types.RoleWindowsDesktop) || !a.hasBuiltinRole(types.RoleDatabase) {
		return nil, trace.AccessDenied("access denied")
	}
	return a.authServer.GenerateWindowsDesktopCert(ctx, req)
}
```

Thoughts?

Reply:
I think that the `GenerateWindowsDesktopCert` should not be shared between Database and desktop access flow. We can add an additional parameter to `GenerateDatabaseCertRequest` to indicate that cert should have Smart card extensions and share this logic between GenerateWindowsDesktopCert and `GenerateDatabaseCert` functions.

Reply:
I think that's a decent compromise.